### PR TITLE
fix Issue 15777 - Premature expansion of overload set in tuples

### PR DESCRIPTION
**Pull request by 9rnsr**

### CODE REVIEW

File: test/runnable/foreach5.d [hunk 1]
```diff
@@ -1085,6 +1087,45 @@ void test14653()
...
+{
+    import imports.test15777a;
+    import imports.test15777b;
+    alias funAB15777 = fun;
+}
+
+void foo15777(alias tpl)()
+{
+    alias seq = AliasSeq!(tpl!());
+    // Make alias of 'overload set' in tuple elements
```
Comment (ZombineDev) [inline test/runnable/foreach5.d]: If I understand correctly, `seq.length` should be 2. Can you please add a static assert for clarity?

Comment (9rnsr) [inline test/runnable/foreach5.d]: in here, `seq.length` is always 1.

Comment (ZombineDev) [inline test/runnable/foreach5.d]: Sorry, I was only thinking about the `funAB15777` case. Can you explain what happens there? Shouldn't there be an ambiguity error as the signatures of both functions `foo` are identical? If the two `foo`s were different, would the seq.length equal 2 (when `tpl` is `funAB15777`)?

Comment (9rnsr) [inline test/runnable/foreach5.d]: When `tpl` is `funAB15777`, seq.length == 1 and seq[0] is an 'overload set". As you think, it contains two function symbols `imports.tet15777a.fun` and `imports.tet15777b.fun` as the candidates, but the overload set is still one symbol. That's why seq.length is 1.

And, the signature ambiguity of those candidates in an overload set is not shown until the overload set is actually called. The detailed spec about the "overload set" is here: http://dlang.org/spec/function#overload-sets

Comment (ZombineDev) [inline test/runnable/foreach5.d]: Thank you very much for the detailed explanations! Now I understand. It maybe helpful to add them as comments in the test.

Comment (9rnsr) [inline test/runnable/foreach5.d]: I don't add such comment, because it's merely a detail of language spec and not related with the bugzilla issue.

**Comment by ibuclaw** [comment]
Rebased in #7700

**Pull request closed without merging**